### PR TITLE
phpstan, composer, easyengine, phpbrew, pickle, wp-cli: add `pour_bottle?` block

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting? -- I don't have `php` installed. (Sorry.)
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Keg-relocation will break the install for anyone not using `/usr/local`
as their prefix on Intel macOS.

### CODE REVIEW

**Comment by carlocab** [comment]
CC @Homebrew/core -- we probably need to do this for anything that installs a `.phar` file. See #76044 for context. FYI.

**Comment by danielnachun** [comment]
It looks `.phar` files are binary files (though not executables) which could theoretically be patched like we plan to do for ELF/MachO files.  I'd be curious to see if they need null padding or if they can use regular string replacement.

**Comment by carlocab** [comment]
I doubt you can do either -- php seems to check the SHA1 to verify that the `.phar` hasn't been modified.

**Comment by danielnachun** [comment]
So maybe we need a `skip_relocation` DSL for files where we can ignore prefix strings?  Also is it the case that keg relocation would work on ARM and Linux in a non-default prefix?

**Comment by carlocab** [comment]
> So maybe we need a `skip_relocation` DSL for files where we can ignore prefix strings?

I've suggested this previously but @MikeMcQuaid wasn't keen on the idea. I think we want to try to be a bit more discerning about text relocation first -- I think we currently relocate all instances `/usr/local` quite aggressively, when we could be a bit more selective about relocation instead.

> Also is it the case that keg relocation would work on ARM and Linux in a non-default prefix?

No, if you attempt to relocate these `.phar` files on ARM or Linux, you'll break these scripts too. It's just that there are no instances of`/opt/homebrew` or `/home/linuxbrew/.linuxbrew` to relocate, so we don't need to worry about keg relocation there.

**Comment by danielnachun** [comment]
I see - so the issue is that these files are not actually including the build prefix at all - they just include `/usr/local`, which happens to be the build prefix for Intel macOS.  I

I think there could be an argument for skipping relocation on specific files, especially in a scenario like this, where the build prefix is not actually being used.  Now that we have 3 different install prefixes, 2 of which are not in the Filesystem Hierarchy Standard, this could make it easier for us to identify cases like this where `/usr/local` is a red herring and doesn't actually need to be relocated.

**Comment by carlocab** [comment]
> I see - so the issue is that these files are not actually including the build prefix at all - they just include `/usr/local`, which happens to be the build prefix for Intel macOS. I

Yes. Note that there's nothing being built here at all; the `install` method just moves some files around. These formulae used to be `bottle :unneeded` before we were bottling everything.

**Comment by MikeMcQuaid** [comment]
> I've suggested this previously but @MikeMcQuaid wasn't keen on the idea. I think we want to try to be a bit more discerning about text relocation first -- I think we currently relocate all instances `/usr/local` quite aggressively, when we could be a bit more selective about relocation instead.

Yes agreed.

> I think there could be an argument for skipping relocation on specific files

Sorry, I don't still like this. It feels like a bit of a hack and something that requires adjusting individual formulae vs. fixing the global behaviour to be more correct. 

Relatedly: how do we know that:
1) these files will never have relocations that are needed (e.g. on bump upgrades)
2) these files will be sufficient for relocations

File: Formula/composer.rb [hunk 1]
```diff
@@ -17,6 +17,14 @@ class Composer < Formula
     sha256 cellar: :any_skip_relocation, mojave:        "4c852f706d4ead18bb1cf79a972783b169fa52ce24c348a1d5dd349db5b6f404"
   end
 
+  # Keg-relocation breaks the formula when it replaces `/usr/local` with a non-default prefix
+  pour_bottle? do
+    on_macos do
+      reason "The bottle needs to be installed into `#{Homebrew::DEFAULT_PREFIX}` on Intel macOS."
```
Comment (MikeMcQuaid) [inline Formula/composer.rb:23]: Didn't we have some sort of default `reason` for this? If it doesn't cover this exact case: can that be added?

Comment (carlocab) [reply Formula/composer.rb:23]: We only have `:clt_installed` as a default reason. We can add one for this, but that'll need to wait for the next release tag to be used. We should probably merge this before then and then come back to clean this up after.

Comment (MikeMcQuaid) [reply Formula/composer.rb:23]: Sounds good 👍🏻

Comment (carlocab) [reply Formula/composer.rb:23]: Homebrew/brew#11742

Comment (carlocab) [reply Formula/composer.rb:23]: #81548

**Pull request closed without merging**